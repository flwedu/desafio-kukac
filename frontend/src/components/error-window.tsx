import { useEffect } from "react";
import useErrorContext from "../hooks/error-context";

export function ErrorWindow() {
  const { error, setError } = useErrorContext();

  useEffect(() => {
    if (error) {
      console.error(error);
    }
  }, []);

  function closeWindow() {
    setError(null);
  }

  return error ? (
    <div className="error-window flex flex-col shadow rounded">
      <div className="header">{error.name}</div>
      <div className="body">
        <div>{error.message}</div>
      </div>
      <button onClick={closeWindow}>Fechar</button>
    </div>
  ) : (
    <></>
  );
}
