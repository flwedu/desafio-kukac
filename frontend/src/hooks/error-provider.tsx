import { createContext, useState } from "react";

type ErrorContextType = {
  error: any;
  setError: (err: any) => void;
};

const ErrorContext = createContext<ErrorContextType | null>(null);

const ErrorContextProvider = ({ children }: { children: JSX.Element }) => {
  const [error, setError] = useState<any>(null);

  return (
    <ErrorContext.Provider value={{ error, setError }}>
      {children}
    </ErrorContext.Provider>
  );
};

export { ErrorContext, ErrorContextProvider };
