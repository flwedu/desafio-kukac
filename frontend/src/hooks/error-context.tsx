import { useContext } from "react";
import { ErrorContext } from "./error-provider";

const useErrorContext = () => {
  const context = useContext(ErrorContext);

  if (!context) {
    throw new Error("ErrorContext is undefined");
  }

  return context;
};

export default useErrorContext;
