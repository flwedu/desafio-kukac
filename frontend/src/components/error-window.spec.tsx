import { render, screen } from "@testing-library/react";
import { ErrorContext } from "../hooks/error-provider";
import { ErrorWindow } from "./error-window";

describe("Error window component tests", () => {
  test("Should not render the component if error = null", () => {
    const error = null;
    const setError = jest.fn();

    const { container } = render(
      <ErrorContext.Provider value={{ error, setError }}>
        <ErrorWindow />
      </ErrorContext.Provider>
    );

    expect.assertions(1);
    expect(container.querySelector("#error-window")).toBeFalsy();
  });

  test("Should render the component", () => {
    const error = new Error("Test Error");
    const setError = jest.fn();

    const { container } = render(
      <ErrorContext.Provider value={{ error, setError }}>
        <ErrorWindow />
      </ErrorContext.Provider>
    );

    const errorWindow = container.querySelector("#error-window");

    expect.assertions(3);
    expect(errorWindow).toBeTruthy();
    expect(screen.getByText("Test Error")).toBeTruthy();
    expect(container).toMatchSnapshot();
  });
});

export {};
