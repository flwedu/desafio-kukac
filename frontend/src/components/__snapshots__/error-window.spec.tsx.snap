// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Error window component tests Should render the component 1`] = `
<div>
  <div
    class="error-window flex flex-col shadow rounded"
    id="error-window"
  >
    <div
      class="header"
    >
      Error
    </div>
    <div
      class="body"
    >
      <div>
        Test Error
      </div>
    </div>
    <button>
      Fechar
    </button>
  </div>
</div>
`;
